mesonbuild/compilers/compilers.py

```diff
@@ -753,7 +753,27 @@ def log_var(var, val):
         return preproc_flags, compile_flags, link_flags
 
     def get_options(self):
-        return {} # build afresh every time
+        opts = {} # build afresh every time
+
+        # Take default values from env variables.
+        (preproc_args, compile_args, link_args) = self.get_args_from_envvars()
+        description = 'Extra arguments passed to the {}'.format(self.get_display_language())
+        opts.update({
+            self.language + '_args': coredata.UserArrayOption(
+                self.language + '_args',
+                description + ' compiler',
+                compile_args, shlex_split=True, user_input=True),
+            self.language + '_link_args': coredata.UserArrayOption(
+                self.language + '_link_args',
+                description + ' linker',
+                link_args, shlex_split=True, user_input=True),
+            self.language + '_preprocess_args': coredata.UserArrayOption(
+                self.language + '_preprocess_args',
+                description + ' preprocessor',
+                preproc_args, shlex_split=True, user_input=True),
+        })
```

Comment:
Hm, ok, I think I'll add a `hidden` flag on UserOption class that would make it skipped when mconf print options, and make it considered unknown option in set_option().

Reply:
No! Any sort of preprocess option must *not* be added. Even a hidden one.

Reply:
We still need to use CPPFLAGS from env. I don't think we should have completely separated mechanism just for that.

Reply:
The most important thing in any application is its internal data model. Leaking legacy things from the system to it should be avoided if at all possible. If the internal and external data models differ, then the pain should be on the code that does the conversion. This makes the usage code simpler, which saves time and effort and reduces total system complexity.

Reply:
> There is no reasonable need for preprocess args, we only need to support CPP args for backwards compatibility but we should not expose it to users.

Well, this is not true. We run some compiler checks that only run the pre-processor, so we need to store those separately.

In the future, once we add support for transparently compiling assembly files with MSVC (just like we do with GCC and clang), we will need to support running them through the pre-processor before being compiled with `ml`. See: https://github.com/centricular/libffi/blob/meson/src/meson.build#L55 for roughly what needs to be done.

I think pretending that pre-processor flags don't exist, or that the pre-processor step is the same as compilation is an abstraction mistake. We shouldn't paint ourselves into a corner here so that when someone comes to us with an unavoidable use-case for having separate pre-processor flags, we say "oh, well, meson can't do that".

Reply:
Regardless of everything else, this PR is about refactoring and cleanup specifically saying "does not include new features". New options are very much new features so they do not belong here, _especially_ in a change with commit message "Move <lang>_args to coredata.compiler_options".

Reply:
That ffi preprocessor thing looks a bit freaky. Is that preprocessor thing something that needs to be done with all (most?) asm files or only those specific ones in ffi?

> Also there is no way to override that from the command line. 

They are put in `c_args`, which you can edit with `meson configure` in the usual way.

> After some thinking, I think that option should be exposed.

Again, whether or not they are exposed, it should _not_ be done in this MR, which is about refactoring without changing functionality.

Reply:
> They are put in c_args, which you can edit with meson configure in the usual way.

Only when compiling, but not for compiler checks. _get_compiler_check_args() uses external_preprocess_args in 'preprocess' mode. That actually seems to be the only place CPPFLAGS is used other than when appended to CFLAGS.

> Again, whether or not they are exposed, it should not be done in this MR, which is about refactoring without changing functionality.

Well, I spent already days working on this, I don't want to refactor again this MR if it's to add it back tomorrow because it is useful after all.

Reply:
Or, what about just removing external_preprocess_args completely since it's used only in one place, and I don't think it makes any difference if we pass CFLAGS to preprocessor. At least all tests still pass.

Reply:
> That ffi preprocessor thing looks a bit freaky. Is that preprocessor thing something that needs to be done with all (most?) asm files or only those specific ones in ffi?

It's needed for asm files that need to be pre-processed (`#include`, `#define`, etc), which is most asm files, if not all.

Yes, it's dumb, but it looks scarier than it is. Instead of running a compile directly, we have an extra step. Should be straightforward to implement in the `ninja` backend (barring meson abstraction issues, since we don't want to generate a custom target unnecessarily), but I haven't investigated how to implement it for the vs backend without using the `custom_target()` abstraction.

Reply:
> Or, what about just removing external_preprocess_args completely since it's used only in one place, and I don't think it makes any difference if we pass CFLAGS to preprocessor. At least all tests still pass.

I suspect we will get warnings about unknown args, but we can try this. I looked at the git history, and I can't find anything that says that this won't work.

Reply:
Is this the only thing blocking this PR or the review is not finished yet ?

Reply:
Other than the one minor thing I just added, this is the only remaining issue. Thanks for working this, it's a nice cleanup, it's always a bit unfortunate that big changes take a long time to review.

Reply:
So what's the status here? @nirbheek made a case that preprocess args would be needed for libffi. I personally think we should add this option because it doesn't cost us anything since the patch is already here, and it gives a way to override CPPFALGS for compiler checks from command line.

Should I just edit the commit msg to state this?

Reply:
Updated this commit to keep preprocessor flags out of compiler_options.